### PR TITLE
Move all the printing to the script.

### PR DESCRIPTION
**Pull request by james-w**
Make the module more functional by leaving all the printing of
the data to the script.

Pretty much all shuffling code around, naming things and giving
functions docstrings.

### CODE REVIEW

File: jenkins/_jenkins.py [hunk 1]
```diff
@@ -91,7 +110,34 @@ def _flatten_builds(builds):
...
+
+def make_subbuild_data_frame(builds):
+    """
+    Make a DataFrame from of sub build information.
+
+    This indexes all builds that are triggered by
+    the top level builds.
+
+    :param Iterable[dict] builds: an iterable of
+        dicts of build information.
```
Comment (jml) [inline jenkins/_jenkins.py]: Is this the same type as used in `make_build_data_frame`?

File: jenkins/_jenkins.py [hunk 2]
```diff
@@ -91,7 +110,34 @@ def _flatten_builds(builds):
             yield thing
 
 
-def make_data_frame(builds):
+def make_build_data_frame(builds):
+    """
+    Make a DataFrame from of top-level build information.
+
+    :param Iterable[dict] builds: an iterable of
+        dicts of build information.
```
Comment (jml) [inline jenkins/_jenkins.py]: :question: Maybe add "as obtained from …"

File: jenkins/_jenkins.py [hunk 3]
```diff
@@ -322,7 +331,16 @@ def get_failing_tests(test_report):
     return list(filter(test_case_failed, list_tests(test_report)))
 
 
-def print_commonly_failing_tests(build_data):
+def analyze_failing_tests(build_data):
+    """
+    Given a DataFrame of build data, analyse which
+    individaul tests are failing the builds.
```
Comment (jml) [inline jenkins/_jenkins.py]: nit: "individual"

File: jenkins/_jenkins.py (hunk 1 above)
Comment (james-w) [inline jenkins/_jenkins.py]: Yes, pretty much. It's possible the other one wouldn't have the sub-build
information.

Any ideas for how to define a type that is "dict with _these_ keys?"

File: jenkins/_jenkins.py [hunk 4]
```diff
@@ -74,6 +74,25 @@ def summarize_build_results(builds):
     return collections.Counter(map(get_build_result, builds))
 
 
+def summarize_weekly_stats(builds):
+    """
+    Summarize the per-week data
+
+    :param Iterable[dict] builds: An iterable of build
+        data dicts.
```
Comment (jml) [inline jenkins/_jenkins.py]: :question: Maybe this should just take the data frame?

**Comment by jml** [comment]
Great, thanks. It's always good to see IO & compute split. Makes it much easier to test, eh?

Can you please make sure any functions in _jenkins that aren't used by the scripts are given an underscore prefix, just to make it clear when reading through the module?

Merge after addressing comments at your discretion.

**Pull request merged**